### PR TITLE
Feature/video api

### PR DESCRIPTION
**Pull request by xal9wiii4ik**
setup django and docker; add video
as a user i want to save my new video
as a server we save new video data in db and upload file to aws, we provide for user get all videos or update or delete

added possibility to do development locally using docker, video api + video soft delete; setup django project; 
added Action for test; update test workflow(env variables getting from github secrets)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
+FROM python:3.9-slim as base
```
Comment (artsiom-streetshares) [inline Dockerfile]: best practise is to specify patch version as well, like python:3.9.1

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+ENV PYTHONDONTWRITEBYTECODE 1
+ENV PYTHONUNBUFFERED 1
+
+RUN apt-get update
+RUN apt-get install --no-install-recommends -y curl build-essential
+
+COPY . .
+
+RUN pip3 install --upgrade pip
```
Comment (artsiom-streetshares) [inline Dockerfile:12]: why use pip? it doesn't support reproducible builds

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+RUN pip3 install -r ./app/requirements.txt
+
+RUN chmod u+x ./app/entrypoint.sh
+
+FROM base as dev
+RUN apt-get install -y git
+RUN pip3 install pre-commit mypy flake8
+
+FROM base as prod
```
Comment (artsiom-streetshares) [inline Dockerfile:22]: nothing after this line?

File: app/api/video/migrations/0001_initial.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    operations = [
+        migrations.CreateModel(
+            name='Video',
+            fields=[
+                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('title', models.CharField(max_length=100)),
+                ('description', models.TextField(max_length=1024)),
+                ('bucket_path', models.CharField(max_length=100)),
+                ('upload_date', models.DateTimeField(auto_now_add=True)),
```
Comment (artsiom-streetshares) [inline app/api/video/migrations/0001_initial.py:21]: the name says date, but it's datetime in fact

File: app/api/video/migrations/0002_video_delete_time_alter_video_bucket_path_and_more.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+# Generated by Django 4.0.2 on 2022-02-27 10:03
```
Comment (artsiom-streetshares) [inline app/api/video/migrations/0002_video_delete_time_alter_video_bucket_path_and_more.py:1]: adhere to one migration at a time policy

File: app/api/video/serializers.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    """
+    Model serializer for model video
+    """
+
+    file = serializers.FileField(required=False, write_only=True)
+
+    class Meta:
+        model = Video
+        fields = '__all__'
+        read_only_fields = ['bucket_path']
```
Comment (artsiom-streetshares) [inline app/api/video/serializers.py]: what about delete_time and upload_date? they read only too

File: app/api/video/services_views.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import boto3
```
Comment (artsiom-streetshares) [inline app/api/video/services_views.py:1]: I don't get this `services_views` name. why not just services?

File: app/api/video/services_views.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    Return:
+        current bucket path
+    """
+
+    s3 = boto3.resource('s3')
+    bucket = s3.Bucket(settings.VIDEOS_BUCKET)
+
+    datetime_now = datetime.now().strftime('%Y-%m-%d:%H-%M-%S.%f')
+    # TODO add username
+    file_path = f'username/{datetime_now}.{file_content_type[0].split("/")[-1]}'
```
Comment (artsiom-streetshares) [inline app/api/video/services_views.py]: extract file_content_type[0].split("/")[-1]} to a variable so it's clear what it is

Comment (artsiom-streetshares) [reply app/api/video/services_views.py]: for the now
1 use utc timezone
2 you can format it right in f-string f'username/{datetime.now():%Y-%m-%d:%H-%M-%S.%f}.
4 if you need custom format store it somewhere in constants
5 if you want ISO format (and you usually want, use https://docs.python.org/3/library/datetime.html#datetime.date.isoformat

File: app/api/video/services_views.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+    file = bucket.put_object(Key=file_path,
+                             Body=file_bytes,
+                             ContentType=file_content_type[0])
+
+    bucket_path = f'https://s3-{settings.BUCKET_REGION}.amazonaws.com/{settings.VIDEOS_BUCKET}/{file.key}'
```
Comment (artsiom-streetshares) [inline app/api/video/services_views.py]: store this template somewhere in a constant, and then do `.format `with parameters

File: app/api/video/tests/test_api/test_api_video.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
+import json
```
Comment (artsiom-streetshares) [inline app/api/video/tests/test_api/test_api_video.py:1]: why not use pytest

File: app/api/video/views.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        # TODO add user
+        serializer.save()
+
+    def destroy(self, request: Request, *args: tp.Any, **kwargs: tp.Any) -> Response:
+        try:
+            video = Video.objects.get(pk=kwargs.get('pk'))
+        except Video.DoesNotExist:
+            return Response(data={"detail": "Not found."}, status=status.HTTP_404_NOT_FOUND)
+        else:
+            video.delete_time = timezone.now()
```
Comment (artsiom-streetshares) [inline app/api/video/views.py:46]: again, use utc timezone

File: app/entrypoint.sh [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    echo "Waiting for psql"
+
+    while ! nc -z $SQL_HOST $SQL_PORT; do
+      sleep 0.1
+    done
+
+    echo "Psql started"
+fi
+
+python app/manage.py makemigrations
```
Comment (artsiom-streetshares) [inline app/entrypoint.sh]: this is not correct. you must create migrations only manually

File: app/entrypoint.sh [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+python app/manage.py migrate
+python app/manage.py collectstatic --noinput
```
Comment (artsiom-streetshares) [inline app/entrypoint.sh:15]: why you need that? your djagno doesn't host any static files

File: app/requirements.txt [hunk 14]
```diff
@@ -0,0 +1,5 @@
+django-cors-headers==3.11.0
+djangorestframework==3.13.1
+psycopg2-binary==2.9.2
+boto3==1.21.8
+fleep
```
Comment (artsiom-streetshares) [inline app/requirements.txt]: version not specified, very bad practice

File: docker-compose-dev.yml [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+  web:
+    restart: always
+    build:
+      context: .
+      dockerfile: Dockerfile
+      target: dev
+    entrypoint: ./app/entrypoint.sh
+    volumes:
+      - ./static_volume/:/staticfiles
+      - ./media_volume/:/mediafiles
```
Comment (artsiom-streetshares) [inline docker-compose-dev.yml:13]: why need these volumes?

File: Dockerfile (hunk 1 above)
Comment (xal9wiii4ik) [reply Dockerfile]: done

File: Dockerfile (hunk 2 above)
Comment (xal9wiii4ik) [reply Dockerfile:12]: we use poetry in flask and mb I chose pip out of habit, if poetry needed i can fix this

File: Dockerfile (hunk 3 above)
Comment (xal9wiii4ik) [reply Dockerfile:22]: yep, i thought I ask you on the call about creating another req file

File: app/api/video/services_views.py (hunk 7 above)
Comment (xal9wiii4ik) [reply app/api/video/services_views.py:1]: i showed that it refers to views and can be services_serializer and etc, i dont know if this is bad practic if that true i will rename

File: app/api/video/migrations/0001_initial.py (hunk 4 above)
Comment (xal9wiii4ik) [reply app/api/video/migrations/0001_initial.py:21]: done

File: app/api/video/migrations/0002_video_delete_time_alter_video_bucket_path_and_more.py (hunk 5 above)
Comment (xal9wiii4ik) [reply app/api/video/migrations/0002_video_delete_time_alter_video_bucket_path_and_more.py:1]: yep, i would like to override base accont model and will need to remove my db and all migrations

File: app/api/video/serializers.py (hunk 6 above)
Comment (xal9wiii4ik) [reply app/api/video/serializers.py]: yes, my bad

Comment (xal9wiii4ik) [reply app/api/video/serializers.py]: done

File: app/api/video/views.py (hunk 11 above)
Comment (xal9wiii4ik) [reply app/api/video/views.py:46]: if i currently remember django timezone use utc by default

File: app/entrypoint.sh (hunk 12 above)
Comment (xal9wiii4ik) [reply app/entrypoint.sh]: ok, done

File: app/entrypoint.sh (hunk 13 above)
Comment (xal9wiii4ik) [reply app/entrypoint.sh:15]: this is for admin panel

File: docker-compose-dev.yml (hunk 15 above)
Comment (xal9wiii4ik) [reply docker-compose-dev.yml:13]: static for admin panel, media can be deleted

Comment (xal9wiii4ik) [reply docker-compose-dev.yml:13]: done

File: app/requirements.txt (hunk 14 above)
Comment (xal9wiii4ik) [reply app/requirements.txt]: done

File: app/api/video/tests/test_api/test_api_video.py (hunk 10 above)
Comment (xal9wiii4ik) [reply app/api/video/tests/test_api/test_api_video.py:1]: in my opinion django and django-rest-framework provide good unit test, but if you think that pytest looks better i will update to pytest

File: app/api/video/services_views.py (hunk 8 above)
Comment (xal9wiii4ik) [reply app/api/video/services_views.py]: i didnt know about the second point, thanks
done

File: app/api/video/services_views.py (hunk 9 above)
Comment (xal9wiii4ik) [reply app/api/video/services_views.py]: done

File: app/api/video/migrations/0001_initial.py [hunk 16]
```diff
@@ -0,0 +1,24 @@
+# Generated by Django 4.0.2 on 2022-02-26 15:04
```
Comment (artsiom-streetshares) [inline app/api/video/migrations/0001_initial.py:1]: you still have many migrations

File: app/api/video/services_views.py (hunk 7 above)
Comment (artsiom-streetshares) [reply app/api/video/services_views.py:1]: I don't see value in such naming. I would not recommend naming like this.

File: app/requirements.txt [hunk 17]
```diff
@@ -0,0 +1,5 @@
+django-cors-headers==3.11.0
```
Comment (artsiom-streetshares) [inline app/requirements.txt:1]: where is django itself?

File: Dockerfile (hunk 2 above)
Comment (artsiom-streetshares) [reply Dockerfile:12]: watch this to see why poetry and pipenv are better than pip https://www.youtube.com/watch?v=aX9GuAC3_aU

**Review by artsiom-streetshares — APPROVED** [review]

**Pull request merged**